src/instrumenter/state_vars.ts

```diff
@@ -0,0 +1,670 @@
+
+        while (lhs.parent instanceof TupleExpression) {
+            const idx = lhs.parent.vOriginalComponents.indexOf(lhs);
+            assert(idx !== -1, ``);
+            idxPath.unshift(idx);
+            lhs = lhs.parent;
+        }
+
+        const assignment = lhs.parent as ASTNode;
+        assert(
+            assignment instanceof Assignment,
+            `Unexpected state var LHS in ${assignment.constructor.name}#${lhs.id} - expected assignment`
+        );
+
+        return [assignment, idxPath];
+    };
+
+    const addStateVarUpdateLocDesc = (
+        node: [Assignment, number[]] | FunctionCall | UnaryOperation,
+        lhs: Expression,
+        rhs: Expression | [Expression, number] | undefined
+    ): void => {
+        const [baseExp, path] = decomposeLHS(lhs);
+
+        // Skip assignments where the base of the LHS is not a direct reference to a state variable
+        if (!isStateVarRef(baseExp)) {
+            return;
+        }
+
+        const stateVarDecl: VariableDeclaration = baseExp.vReferencedDeclaration as VariableDeclaration;
+
+        res.push([node, stateVarDecl, path, rhs]);
+    };
+
+    for (const unit of units) {
+        for (const [lhs, rhs] of getAssignments(unit)) {
+            // Assignments to struct constructor fields - ignore
+            if (lhs instanceof Array) {
+                continue;
+            }
+
+            if (lhs instanceof VariableDeclaration) {
+                // Local variable/function arg/function return - skip
+                if (!lhs.stateVariable) {
+                    continue;
+                }
+
+                assert(
+                    rhs instanceof Expression,
+                    `RHS cannot be a tuple/function with multiple returns.`
+                );
+                // State variable inline initializer
+                res.push([lhs, lhs, [], rhs]);
+                continue;
+            }
+
+            const [baseExp, path] = decomposeLHS(lhs);
+
+            // Skip assignments where the base of the LHS is not a direct reference to a state variable
+            if (!isStateVarRef(baseExp)) {
+                continue;
+            }
+
+            const stateVarDecl: VariableDeclaration = baseExp.vReferencedDeclaration as VariableDeclaration;
+
+            res.push([getLHSAssignmentNode(lhs), stateVarDecl, path, rhs]);
+        }
+
+        // Find .push() and pop()
+        for (const candidate of unit.getChildrenBySelector(
+            (node) =>
+                node instanceof FunctionCall &&
+                node.vFunctionCallType === ExternalReferenceType.Builtin &&
+                (node.vFunctionName === "push" || node.vFunctionName === "pop")
+        )) {
+            const funCall = candidate as FunctionCall;
+
+            addStateVarUpdateLocDesc(
+                funCall,
+                (funCall.vExpression as MemberAccess).vExpression,
+                undefined
+            );
+        }
+
+        // Find all deletes
```

Comment:
We better remove this note (as this is explicit in the following code), or expand it a bit further if there any caveats.

Reply:
Fixed in 35d34c0.